### PR TITLE
bpo-35365: Use a wchar_t* buffer in the code page decoder.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue35365

### CODE REVIEW

**Review by vstinner — COMMENTED** [review]
The allocated size is used the "current position", whereas the code uses overallocation. It seems like your code doesn't introduce the confusion (I wrote this code if I recall properly), but it might be worth it to take this change as an opportunity to clarify what is the allocated size and what is the current position. Maybe use a small structure which contains (buffer, allocated, pos)? The structure may only be updated at function exit for performance, or before calling a function using this structure, like widechar_resize().

Currently, the code works because decode_code_page_errors() resize the overallocated buffer at exit.

In practice, decode_code_page_strict() and decode_code_page_errors() are only be called once, so it should be fine. They are only called multiple times for strings longer than INT_MAX bytes :-)

File: Objects/unicodeobject.c [hunk 1]
```diff
@@ -7314,14 +7308,10 @@ decode_code_page_errors(UINT code_page,
...
-    *out = 0;
-
-    /* Extend unicode object */
+    /* Shrink the buffer */
     outsize = out - startout;
-    assert(outsize <= PyUnicode_WSTR_LENGTH(*v));
-    if (unicode_resize(v, outsize) < 0)
-        goto error;
+    assert(outsize <= *bufsize);
+    *bufsize = outsize;
```
Comment (vstinner) [inline Objects/unicodeobject.c]: Example of place where it becomes unclear if bufsize is the allocation size or the current position :-/

**Comment by vstinner** [comment]
Thanks for doing that, I wanted to do it since I wrote the code a few years ago :-)

**Comment by serhiy-storchaka** [comment]
I thought about this before. But since decode_code_page_strict() and decode_code_page_errors() are only be called once in common case, I prefer to keep the current allocation strategy for simplicity. Changing it is not trivial, and changes should be larger for making it efficient.

When I tried to rewrite this code, I got a suspense that it does not work correctly for large input (>INT_MAX bytes). I'll try to fix it before merging this PR.

File: Objects/unicodeobject.c [hunk 2]
```diff
@@ -7160,21 +7175,12 @@ decode_code_page_strict(UINT code_page,
...
-        out = PyUnicode_AS_UNICODE(*v);
-    }
-    else {
-        /* Extend unicode object */
-        Py_ssize_t n = PyUnicode_GET_SIZE(*v);
-        if (unicode_resize(v, n + outsize) < 0)
-            return -1;
-        out = PyUnicode_AS_UNICODE(*v) + n;
+    /* Extend a wchar_t* buffer */
+    Py_ssize_t n = *bufsize;
```
Comment (vstinner) [inline Objects/unicodeobject.c]: Can you please add a comment explaining that buf is not (must not?) be overallocated, so bufsize is the current length? Something like:

```suggestion
    Py_ssize_t n = *bufsize;   /* Get the current length */
```

File: Objects/unicodeobject.c [hunk 3]
```diff
@@ -7238,29 +7245,16 @@ decode_code_page_errors(UINT code_page,
...
-            PyErr_NoMemory();
-            goto error;
-        }
-        /* FIXME: don't use _PyUnicode_New(), but allocate a wchar_t* buffer */
-        *v = (PyObject*)_PyUnicode_New(size * Py_ARRAY_LENGTH(buffer));
-        if (*v == NULL)
-            goto error;
-        out = PyUnicode_AS_UNICODE(*v);
+    /* Extend a wchar_t* buffer */
+    Py_ssize_t n = *bufsize;
```
Comment (vstinner) [inline Objects/unicodeobject.c]: Same remark:

```suggestion
    Py_ssize_t n = *bufsize;   /* Get the current length */
```

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
Azure Pipelines PR failed because of https://bugs.python.org/issue35386 : the failure can be ignored.

**Pull request merged**